File: pulsar/java/src/main/java/com/examples/hello/pulsar/PulsarExecutors.java
package com.examples.hello.pulsar;

import io.micronaut.context.annotation.Factory;
import jakarta.inject.Named;
import jakarta.inject.Singleton;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Factory
public class PulsarExecutors {
  public static final String CONSUMER = "consumer";
  public static final String PRODUCER = "producer";
  public static final String FUNCTIONS = "functions";

  @Singleton
  @Named(PRODUCER)
  public ExecutorService producer() { return Executors.newFixedThreadPool(5); }

  @Singleton
  @Named(CONSUMER)
  public ExecutorService consumer() {
    return Executors.newFixedThreadPool(5);
  }

  @Singleton
  @Named(FUNCTIONS)
  public ExecutorService functions() {
    return Executors.newFixedThreadPool(5);
  }

}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/PulsarExecutors.kt
package com.example.hello.pulsar

import io.micronaut.context.annotation.Factory
import jakarta.inject.Named
import jakarta.inject.Singleton
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@Factory
class PulsarExecutors {
    @Singleton
    @Named(PRODUCER)
    fun producer(): ExecutorService {
        return Executors.newFixedThreadPool(5)
    }

    @Singleton
    @Named(CONSUMER)
    fun consumer(): ExecutorService {
        return Executors.newFixedThreadPool(5)
    }

    @Singleton
    @Named(FUNCTIONS)
    fun functions(): ExecutorService {
        return Executors.newFixedThreadPool(5)
    }

    companion object {
        const val CONSUMER = "consumer"
        const val PRODUCER = "producer"
        const val FUNCTIONS = "functions"
    }
}
